includes/html/print-alert-rules.php

```diff
@@ -175,23 +220,123 @@
 
     $device_count = dbFetchCell('SELECT COUNT(*) FROM alert_device_map WHERE rule_id=?', [$rule['id']]);
     $group_count = dbFetchCell('SELECT COUNT(*) FROM alert_group_map WHERE rule_id=?', [$rule['id']]);
-    if ($device_count && $group_count) {
-        $popover_msg = 'Restricted rule';
-        $icon_indicator = 'fa fa-connectdevelop fa-fw text-primary';
-    } elseif ($device_count) {
-        $popover_msg = 'Device restricted rule';
+    if ($device_count) {
+        $popover_msg = 'Device restricted rule #' . $rule['id'];
         $icon_indicator = 'fa fa-server fa-fw text-primary';
-    } elseif ($group_count) {
-        $popover_msg = 'Group restricted rule';
+    }
+    if ($group_count) {
+        $popover_msg = 'Group restricted rule #' . $rule['id'];
         $icon_indicator = 'fa fa-th fa-fw text-primary';
-    } else {
-        $popover_msg = 'Global alert rule';
+    }
+    if ($device_count && $group_count) {
+        $popover_msg = 'Device and Group restricted rule #' . $rule['id'];
+        $icon_indicator = 'fa fa-connectdevelop fa-fw text-primary';
+    }
+    if (!$device_count && !$group_count) {
+        $popover_msg = 'Global alert rule #' . $rule['id'];
         $icon_indicator = 'fa fa-globe fa-fw text-success';
     }
 
-    echo "<tr class='".$extra."' id='row_".$rule['id']."'>";
-    echo "<td><i>#".((int) $rule['id'])."</i><br /><i class=\"$icon_indicator\"></i></td>";
+    echo "<tr class='".$extra."' id='rule_id_".$rule['id']."'>";
+
+    // Type
+
+    echo "<td colspan=\"2\"><div data-toggle='popover' data-placement='top' data-content=\"$popover_msg\" class=\"$icon_indicator\"></div></td>";
+
+    // Name
+
     echo '<td>'.$rule['name'].'</td>';
+
+    // Devices (and Groups)
+
+    if ($rule['invert_map'] == 0) {
+        $groups_msg = 'Only devices in this group.';
+        $devices_msg = 'Only this device.';
+        $except_device_or_group = null;
+    }
+
+    if ($rule['invert_map'] == 1) {
+        $devices_msg = 'All devices EXCEPT this device. ';
+        $groups_msg = 'All devices EXCEPT this group.';
+        $except_device_or_group = '<strong><em>EXCEPT</em></strong> ';
+    }
+
+    $device_and_group_queries=array();
+    if ($group_count) {
+        $device_and_group_queries[] = 'SELECT device_groups.name, device_groups.id FROM alert_group_map, device_groups WHERE alert_group_map.rule_id=? and alert_group_map.group_id = device_groups.id ORDER BY name';
+    }
+
+    if ($device_count) {
+        $device_and_group_queries[] = 'SELECT devices.device_id,devices.hostname FROM alert_device_map, devices WHERE alert_device_map.rule_id=? and alert_device_map.device_id = devices.device_id ORDER BY hostname';
+    }
+
+    $devices_and_groups_popover='left';
+
```

Comment:
line 264 - 288
not the best coding style
you could avoid `$device_and_group_queries` array and `preg_match` command
when handling line 278-286 in `if($group_count)` and `if($device_count)` statement seperate

Reply:
I agree.  That's the way I'd originally written it.    As you can see here, 

https://github.com/librenms/librenms/pull/11115/commits/6338e1c7ec4529455d1fbea843c9faebc79b9161

However, codeclimate failed that approach due to DRY.  The way it's written now passes all codeclimate checks.  Meh.  Humans versus algorithms.  I'd be happy to change it back to what we see as a better style, as long as the code gets merged (and someone tunes up codeclimate a bit better for DRY).  There's tons of reuse in LibreNMS.

Reply:
Code Climate is not a required for Merging. It's primary a styleguide hint.

Reply:
yeah, change it back :+1: 

If Device/DeviceGroups mouseover is also changed as i wrote somewhere before, i'm happy to approve your PR for merging.

Reply:
But I don't like red x's.  :)

I do like the way I had `if ($group_count)` && `if $(device_count)` originally, though.  Much easier to userstand IMHO.  If everyone's fine with that approach then so shall it be.

Something's wrong with scrutinizer now, too?

Reply:
commit the easier way, maybe it was only a run problem.
Sometimes happens :/

Reply:
yes, and laravel migrations give red x's too.

codeclimate needs some tuning ...

I put the code back to the way we both prefer.